src/engine.rs

```diff
@@ -1126,4 +1151,112 @@ mod tests {
             });
         }
     }
+
+    #[test]
+    fn test_dump_file_or_directory() {
+        let dir = tempfile::Builder::new()
+            .prefix("test_dump_file_or_directory")
+            .tempdir()
+            .unwrap();
+        let mut file_id = FileId {
+            queue: LogQueue::Rewrite,
+            seq: 7,
+        };
+        let builder = Arc::new(DefaultFileBuilder);
+        let entry_data = vec![b'x'; 1024];
+
+        let mut batches = vec![vec![LogBatch::default()]];
+        let mut batch = LogBatch::default();
+        batch
+            .add_entries::<Entry>(7, &generate_entries(1, 11, Some(&entry_data)))
+            .unwrap();
+        batch.add_command(7, Command::Clean);
+        batch.put(7, b"key".to_vec(), b"value".to_vec());
+        batch.delete(7, b"key2".to_vec());
+        batches.push(vec![batch.clone()]);
+        let mut batch2 = LogBatch::default();
+        batch2.put(8, b"key3".to_vec(), b"value".to_vec());
+        batch2
+            .add_entries::<Entry>(8, &generate_entries(5, 15, Some(&entry_data)))
+            .unwrap();
+        batches.push(vec![batch, batch2]);
+
+        for bs in batches.iter_mut() {
+            let file_path = file_id.build_file_path(dir.path());
+            // Write a file.
+            let mut writer =
+                build_file_writer(builder.as_ref(), &file_path, true /*create*/).unwrap();
+            for batch in bs.iter_mut() {
+                let offset = writer.offset() as u64;
+                let len = batch.finish_populate(1 /*compression_threshold*/).unwrap();
+                writer
+                    .write(batch.encoded_bytes(), 0 /*target_file_hint*/)
+                    .unwrap();
+                batch.finish_write(FileBlockHandle {
+                    id: file_id,
+                    offset,
+                    len,
+                });
+            }
+            writer.close().unwrap();
+            file_id.seq += 1;
+        }
+
+        //dump dir with raft groups. 8 element with raft groups 7 and 2 elements with raft groups 8
+        let raft_groups_ids = &[7u64; 1];
+        let dump_it = Engine::dump(dir.path()).unwrap();
+        let mut total = 0;
+        for item in dump_it {
+            if raft_groups_ids.is_empty() || raft_groups_ids.contains(&item.raft_group_id) {
+                total += 1;
+            }
+        }
+        assert!(total == 8);
```

Comment:
No need to involve `raft_groups_ids` here as well. Verify the total count is enough.

Reply:
🆗